File: 4_step_4_binary_search/05_smallest_divisor_given_a_threshold.cpp
#include <bits/stdc++.h>
using namespace std;

int getMaxElement(vector<int> &arr);
int getCurrentSum(vector<int> &arr, int &num);
int findSmallestDivisor(vector<int> &arr, int &threshold);

int main()
{
    // vector<int> arr = {1, 2, 5, 9};
    // int threshold = 6;
    vector<int> arr = {44,22,33,11,1};
    int threshold = 5;

    int result = findSmallestDivisor(arr, threshold);
    cout << result << endl;

    return 0;
}

int getMaxElement(vector<int> &arr)
{
    int maxElement = INT_MIN;

    for (int i = 0; i < arr.size(); i++)
    {
        if (arr[i] > maxElement)
        {
            maxElement = arr[i];
        }
    }

    return maxElement;
}

int getCurrentSum(vector<int> &arr, int &num)
{
    int currentSum = 0;

    for (int i = 0; i < arr.size(); i++)
    {
        currentSum += (arr[i] + num - 1) / num;
    }

    return currentSum;
}

int findSmallestDivisor(vector<int> &arr, int &threshold)
{
    int low = 1, high = getMaxElement(arr), ans = 0;

    while (low <= high)
    {
        int mid = low + (high - low) / 2;
        int currentSum = getCurrentSum(arr, mid);
        if (currentSum <= threshold)
        {
            ans = mid;
            high = mid - 1;
        }
        else
        {
            low = mid + 1;
        }
    }

    return ans;
}

File: 4_step_4_binary_search/05_smallest_divisor_given_a_threshold.js
function getMaxElement(bloomDay) {
  let maxElement = bloomDay[0];

  for (let i = 0; i < bloomDay.length; i++) {
    if (bloomDay[i] > maxElement) {
      maxElement = bloomDay[i];
    }
  }

  return maxElement;
}

function getCurrentSum(bloomDay, x) {
  let currentSum = 0;

  for (let i = 0; i < bloomDay.length; i++) {
    currentSum += Math.ceil(bloomDay[i]/x);
  }

  return currentSum;
}

function getSmallestDivisor(bloomDay, threshold) {
  let low = 1;
  let high = getMaxElement(bloomDay);
  let ans = -1;

  while (low <= high) {
    let mid = Math.floor(low + (high - low) / 2);
    let currentSum = getCurrentSum(bloomDay, mid);

    if (currentSum <= threshold) {
      ans = mid;
      high = mid - 1;
    } else {
      low = mid + 1;
    }
  }

  return ans;
}

const bloomDay= [44,22,33,11,1]
const threshold = 5;
// const bloomDay= [1,2,5,9]
// const threshold = 6;
let result = getSmallestDivisor(bloomDay, threshold)
console.log(result)